pep-9999.rst

```diff
@@ -0,0 +1,252 @@
+    T_contra = TypeVar("T_contra", contravariant=True)
+    
+    class SimpleFunction(Protocol[T_contra, T_co]):
+        def __call__(self, x: T_contra) -> T_co: ...
+
+But type variables can also be used outside of the scope of generic 
+types, where the concept of variance does not apply, e.g.:
+::
+
+    def first(things: Iterable[T]) -> T:
+        return next(iter(things))
+
+
+This example is valid, but using ``T_co`` or ``T_contra`` instead of 
+``T`` here does not make sense, and type checkers do not allow is.
+
+In ``SimpleFunction``, ``T_co`` or ``T_contra`` are used to annotate
+its method as well. This is required because the variance is associated 
+with the individual type variables. But the method signature is unrelated 
+to the variance of the type parameters of its generic type, 
+``SimpleFunction`` in this example. 
+
+Because variance is restricted to the type parameters of generic types, 
+it makes more sense to specify the variance of type parameters directly
+where the generic type is defined, removing the need to specify it on 
+the type variables themselves. 
+
+
+Proposal
+========
+
+The variance of the type parameters of a generic class can be specified
+with the ``+`` and ``-`` prefix operators on their respective type
+variables within the type parameter list of ``typing.Generic``, 
+``typing.Protocol``, or the generic base class. Instead of
+::
+
+    T_co = typing.TypeVar('T_co', covariant=True)
+    T_contra = typing.TypeVar('T_contra', contravariant=True)
+
+    class SupportsInvertOld(typing.Protocol[T_co]):
+        def __invert__(self) -> T_co: ...
+
+    class SupportsPartialOrderOld(typing.Protocol[T_contra]):
+        def __le__(self, other: T_contra): ...
+
+
+The new syntax uses the ``+`` and ``-`` prefix operators to specify
+variadic type parameters in the same place where the generic type is
+declared:
+::
+
+    T = TypeVar("T")
+    
+    # covariant
+    class SupportsInvert(typing.Protocol[+T]):
+        def __invert__(self) -> T: ...
+
+    # contravariant
+    class SupportsPartialOrder(typing.Protocol[-T]):
+        def __le__(self, other: T): ...
+
+This syntax is inspired by Scala programming language [1]_.
+
+
+
+Specification
+=============
+
+
+Valid use locations
+-------------------
+
+
+The new type variable variance syntax can only be used within the type
+parameter list of ``typing.Generic``, ``typing.Protocol``, or a generic 
+base class. 
+
+The ``+`` and ``-`` prefix operators can be used on ``typing.TypeVar``
+and ``typing.ParamSpec`` instances, that are now already variadic, i.e.
+its ``__covariant__`` and ``__contravariant__`` attributes must be 
+``False``.
+
+When the same type variable is used on multiple generic base classes,
+they must share the same variance, e.g.
+::
+    
+    from typing import TypeVar, Callable, Container, Iterable, Protocol
+    
+    class LinkedList(Iterable[+T], Container[+T]): ...
+    class EventListener(Callable[[-T], None], Protocol[-T]): ...
```

Comment:
I don't think you could use `Callable` in class bases. Consider changing to another type with contravariant parameters.

Reply:
Why not? It's an alias of `collections.abc.Callable`: https://docs.python.org/3/library/collections.abc.html#collections.abc.Callable

Reply:
I might be behind on this, but PEP-585 doesn't say that `collections.abc.Callable` can be indexed in the same way as the `Callable` special-form. 

The typing docs, thought, do [say](https://docs.python.org/3.10/library/typing.html#typing.Callable) that

> A plain `Callable` is equivalent to `Callable[..., Any]`, and in turn to `collections.abc.Callable`.

which makes me think that you cannot use `Callable` in bases, but then `typing.Callable` is deprecated since 3.9 :man_shrugging:... 

As an extra data-point mypy 0.910 rejects your example module the new +/- syntax.

Reply:
You're right @superbobry , not only is `Callable` not a protocol, it's not even a valid base class apparently 🤔 ,  I assumed that it was a protocol with a `__call__`  method that has covariant arg and kwarg types, and a contravariant return type, but I suppose that it was wishful thinking.